src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+const THUNDERCLOUD_MILLIS = 18000
+
+const AFUIBUFFMAXSTACK = 3
+
+const ISSUE_SEVERITY_TIERS = {
+	1: SEVERITY.MINOR,
+	3: SEVERITY.MEDIUM,
+	5: SEVERITY.MAJOR,
+}
+
+const CYCLE_ENDPOINTS = [
+	ACTIONS.BLIZZARD_III.id,
+	ACTIONS.TRANSPOSE.id,
+	ACTIONS.FREEZE.id,
+]
+
+// This is feelycraft at the moment. Rotations shorter than this won't be processed for errors.
+const MIN_ROTATION_LENGTH = 3
+
+/**
+ * Error type codes, higher values indicate higher priority errors. If you add more, adjust the IDs to ensure correct priorities.
+ * Only the highest priority error will be displayed in the 'Reason' column.
+ * NOTE: Cycles with values below ERROR_CODES.SHORT will be filtered out of the RotationTable display
+ * unless the DEBUG_SHOW_ALL_CYCLES variable is set to true
+ */
+const ERROR_CODES = {
+	NONE: 0,
+	FINAL_OR_DOWNTIME: 1,
+	SHORT: 2,
+	MISSING_FIRE4S: 10, // These two errors are lower priority since they can be determined by looking at the
+	MISSING_DESPAIRS: 15, // target columns in the table, so we want to tell players about other errors first
+	MANAFONT_BEFORE_DESPAIR: 30,
+	EXTRA_T3: 49, // Extra T3 and Extra F1 are *very* similar in terms of per-GCD potency loss
+	EXTRA_F1: 50, // These two codes should stay close to each other
+	NO_FIRE_SPELLS: 75,
+	DROPPED_ENOCHIAN: 100,
+}
+
+/**
+ * If you add an error code above, give it a message here too, even ones filtered out normally
+ * so they show something in debug mode.
+ */
+const ERROR_MESSAGES = {
+	[ERROR_CODES.NONE]: <Trans id="blm.rotation-watchdog.error-messages.none">No errors</Trans>,
+	[ERROR_CODES.FINAL_OR_DOWNTIME]: <Trans id="blm.rotation-watchdog.error-messages.final-or-downtime">Ended with downtime, or last cycle</Trans>,
+	[ERROR_CODES.SHORT]: <Trans id="blm.rotation-watchdog.error-messages.short-rotation">Too short, won't process</Trans>,
+	[ERROR_CODES.DROPPED_ENOCHIAN]: <Trans id="blm.rotation-watchdog.error-messages.dropped-enochian">Dropped <ActionLink {...ACTIONS.ENOCHIAN}/></Trans>,
+	[ERROR_CODES.MISSING_FIRE4S]: <Trans id="blm.rotation-watchdog.error-messages.missing-fire4s">Missing <ActionLink {...ACTIONS.FIRE_IV}/>s</Trans>,
+	[ERROR_CODES.EXTRA_T3]: <Trans id="blm.rotation-watchdog.error-messages.extra-t3">Extra <ActionLink {...ACTIONS.THUNDER_III}/>s</Trans>,
+	[ERROR_CODES.MANAFONT_BEFORE_DESPAIR]: <Trans id="blm.rotation-watchdog.error-messages.manafont-before-despair"><ActionLink {...ACTIONS.MANAFONT}/> used before <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.MISSING_DESPAIRS]: <Trans id="blm.rotation-watchdog.error-messages.missing-despair">Rotation didn't include <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.NO_FIRE_SPELLS]: <Trans id="blm.rotation-watchdog.error-messages.no-fire-spells">Rotation included no Fire spells</Trans>,
+	[ERROR_CODES.EXTRA_F1]: <Trans id="blm.rotation-watchdog.error-messages.extra-f1">Extra <ActionLink {...ACTIONS.FIRE_I}/></Trans>,
+}
+
+class Cycle {
+	casts: any[] = []
```

Comment:
Avoid `any`. If there's absolutely no way to type it yet, use `TODO` and comment what it should be.

Reply:
agh, I thought I caught all of those. Fix't